zerver/actions/message_send.py

```diff
@@ -1476,6 +1494,55 @@ def check_sender_can_access_recipients(
         raise JsonableError(_("You do not have permission to access some of the recipients."))
 
 
+def get_recipients_for_user_creation_events(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile]
+) -> Dict[UserProfile, Set[int]]:
+    """
+    This function returns a dictionary with data about which users would
+    receive stream creation events due to gaining access to a user.
+    The key of the dictionary is a user object and the value is a set of
+    user_ids that would gain access to that user.
+    """
+    recipients_for_user_creation_events: Dict[UserProfile, Set[int]] = defaultdict(set)
+
+    if realm.can_access_all_users_group.name == SystemGroups.EVERYONE:
+        return recipients_for_user_creation_events
+
+    if len(user_profiles) == 1:
+        if not check_can_access_user(sender, user_profiles[0]):
+            recipients_for_user_creation_events[sender].add(user_profiles[0].id)
+        return recipients_for_user_creation_events
+
+    guest_recipients = [user for user in user_profiles if user.is_guest]
+    if len(guest_recipients) == 0:
+        return recipients_for_user_creation_events
```

Comment:
Let's add a comment noting that this block relies on the assumption that `realm.can_access_all_users_group` contains all `MEMBERS`. Maybe:

> If none of the users in the direct message conversation are guests, then there is no possible `realm.can_access_all_users_group` policy that would mean sending this message changes any user's user access to other users.

We might also want to consider putting this check before the `realm.can_access_all_users_group` one, since that might save a database query when `realm.can_access_all_users_group` is not prefetched?

Reply:
Applied.